kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/FilterTlsIT.java

```diff
@@ -0,0 +1,265 @@
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import io.netty.handler.ssl.ClientAuth;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.SslProvider;
+
+import io.kroxylicious.net.TlsServer;
+import io.kroxylicious.proxy.config.FilterDefinition;
+import io.kroxylicious.proxy.config.FilterDefinitionBuilder;
+import io.kroxylicious.proxy.config.tls.FilePassword;
+import io.kroxylicious.proxy.config.tls.FilePasswordFilePath;
+import io.kroxylicious.proxy.config.tls.InlinePassword;
+import io.kroxylicious.proxy.config.tls.InsecureTls;
+import io.kroxylicious.proxy.config.tls.KeyStore;
+import io.kroxylicious.proxy.config.tls.Tls;
+import io.kroxylicious.proxy.config.tls.TrustStore;
+import io.kroxylicious.proxy.filter.FilterFactoryContext;
+import io.kroxylicious.proxy.filter.TlsTestFilter;
+import io.kroxylicious.proxy.filter.TlsTestFilter.Outcome;
+import io.kroxylicious.proxy.filter.TlsTestFilterFactory;
+import io.kroxylicious.test.Request;
+import io.kroxylicious.test.ResponsePayload;
+import io.kroxylicious.testing.kafka.common.KeytoolCertificateGenerator;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+import static io.kroxylicious.UnknownTaggedFields.unknownTaggedFieldsToStrings;
+import static io.kroxylicious.test.tester.KroxyliciousConfigUtils.proxy;
+import static io.kroxylicious.test.tester.KroxyliciousTesters.mockKafkaKroxyliciousTester;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Users can use the FilterFactoryContext to obtain an {@link javax.net.ssl.SSLContext} given some
+ * re-usable Tls configuration. This test checks that the context offered to the client behaves
+ * as we hope, enabling a client to connect successfully to a trusted TLS server when we supply the
+ * trust material in various ways. Also, we check that we can supply key material for mutual TLS auth
+ * where the server checks that it trusts the client.
+ *
+ * Note that we discriminate between the FilterFactoryContext passed to {@link io.kroxylicious.proxy.filter.FilterFactory#initialize(FilterFactoryContext, Object)}
+ * and the one passed to {@link io.kroxylicious.proxy.filter.FilterFactory#createFilter(FilterFactoryContext, Object)} as
+ * there is no guarantee they are the same context with the same behaviour.
+ */
+class FilterTlsIT {
+    @TempDir
+    Path tempDir;
+    @TempDir
+    Path tempDir2;
+
+    record Keys(Path keyStore, Path trustStore, Path certfilePath, String password, String keyStoreType, String trustStoreType) {}
+
+    record TestConfig(String name, Function<Keys, Tls> filterTlsConfig, Outcome outcome) {
+
+        @Override
+        public String toString() {
+            return name;
+        }
+    }
+
+    private static List<TestConfig> tlsConfigurations() {
```

Comment:
Oh, I didn't know you could do this.  I'd always use a Stream<Arguments> even when I have only one parameter to pass.